### ISSUE TITLE
Why layer norm here before sublayer and addition?

### ISSUE THREAD

**Issue by fbioki** [tensorflow/models#9789]
Hi,
code reference: https://github.com/tensorflow/models/blob/bee6a47121f16e5903463cf3537a3e09f6be42e0/official/nlp/transformer/transformer.py#L367

**Why layer norm here before sublayer and addition?** According to paper, Attention is all you need, "We employ a residual connection [11] around each of the two sub-layers, followed by layer normalization [1]. That is, the output of each sub-layer is LayerNorm(x + Sublayer(x)), where Sublayer(x) is the function implemented by the sub-layer itself". The layer normalization is the end process after self-attention and add.

**Comment by lehougoogle** [comment]
Hi fbioki,

The original transformer uses Post-norm. But actually Pre-norm [1, 2] works better.

[1] He, Kaiming, et al. "Identity mappings in deep residual networks." European conference on computer vision. Springer, Cham, 2016.
[2] Wang, Qiang, et al. "Learning deep transformer models for machine translation." arXiv preprint arXiv:1906.01787 (2019).

**Comment by fbioki** [comment]
Thanks

**Comment by saberkun** [comment]
Please checkout our new implementation that has both paths: https://github.com/tensorflow/models/blob/master/official/nlp/keras_nlp/layers/transformer_encoder_block.py#L104
